[SecuritySolution][EntityAnalytics][PrivMon] Disable privmon engine API endpoint

abhishekbhatia1710 opened the pull request:
## Summary

A [document](https://docs.google.com/document/d/1bumJbnJZyyjyXJeYZGmsgU867geEuEECOu46GKMWYss) describes the thought process while creating the `disable` API endpoint for Privileged User monitoring.

## Summary of the document : 

Based on a comparison with the existing `disable`, `delete`, and `stop` endpoints in the `risk_engine`, `risk_score`, and `entity_store` modules ,and considering that no transforms are currently involved in PrivMon, this PR implements only the disable endpoint for PrivMon . The stop endpoint has been intentionally not chosen (“parked for future use”) and can be introduced later if transforms become relevant in PrivMon, ensuring alignment with the overall endpoint strategy across these modules.

The `disable` endpoint performs : 
1. Removing the privmon task
2. Updating the privmon saved object with `status: disabled`

And reverts all the changes to its earlier state in case of any failure.

## Endpoint details

`POST /api/entity_analytics/monitoring/engine/disable`

Response : 

### If the engine is already disabled
```
{
    "status": "disabled",
    "error": null
}
```

<img width="1342" height="224" alt="image" src="https://github.com/user-attachments/assets/eb9f1d00-741e-47ba-8fc4-a9e58d1df59f" />

### If the engine is running 

{
    "status": "disabled",
    "error": null
}

<img width="1348" height="300" alt="image" src="https://github.com/user-attachments/assets/4d9f2190-ae2b-4375-b320-874a5776672d" />

### If the disable operation fails
```
{
    "status": "started",
    "error": [
        "Failed to disable Privileged Monitoring Engine: Task Manager is not available"
    ]
}
```
<img width="1348" height="178" alt="image" src="https://github.com/user-attachments/assets/ff99892b-f367-4cd6-ac88-85c5a3594444" />


### Testing Steps 
1. Pull down this PR.
2. Enable the priv mon engine using dev console. `POST kbn:/api/entity_analytics/monitoring/engine/init`
3. Have fun playing around with this endpoint.

### Test cases result

<img width="1860" height="701" alt="image" src="https://github.com/user-attachments/assets/8d5acd9a-7e91-4582-b87f-8f2ff17da952" />




### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.





<!--ONMERGE {"backportTargets":["9.1"]} ONMERGE-->

abhishekbhatia1710 committed:
Inital commit for the disable API endpoint

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/317685)
* Commit: b5c76809e1c30e290f204c3b1696edeba364b5c4

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/317685#0197f42d-5df0-4e93-84c3-2a8173d14ec1)

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317685","number":317685,"commit":"b5c76809e1c30e290f204c3b1696edeba364b5c4"}],"number":317685}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Changed EngineStatus to PrivilegedMonitoringEngineStatus

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

hop-dev reviewed (changes requested):
Great work here! one point to maybe discuss wider.

abhishekbhatia1710 committed:
Merge branch 'main' into ea-12990-disable-privmon

abhishekbhatia1710 committed:
Removed
- the code and tests for "revert to earlier engine state".
- the code to disable each data source

abhishekbhatia1710 committed:
Merge branch 'main' into ea-12990-disable-privmon

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/319055)
* Commit: 70ed4bafc4353364bcdbcb3b95e2a3b35c9eeabd

### Failed CI Steps
* [FTR Configs #22](https://buildkite.com/elastic/kibana-pull-request/builds/319055#01980ca0-2c49-4c30-90f7-bf7746780a2d)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/319055#01980ca0-2c49-4c30-90f7-bf7746780a2d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/319055/jobs/01980ca0-2c49-4c30-90f7-bf7746780a2d/artifacts/01980cc4-a1f8-485f-9c07-c5aa209e3761) FTR Configs #<span></span>22 / Visualizations - Group 2 lens app - Agg based Vis Open in Lens Table should convert percentage column

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #318540](https://buildkite.com/elastic/kibana-pull-request/builds/318540) succeeded 3138f96ea7b0f0b410e569a9a84e0f30de06517d
* :yellow_heart: [Build #318023](https://buildkite.com/elastic/kibana-pull-request/builds/318023) was flaky 36728a5e23c719d6c24c274aa05704eaee5817f3
* :broken_heart: [Build #317685](https://buildkite.com/elastic/kibana-pull-request/builds/317685) failed b5c76809e1c30e290f204c3b1696edeba364b5c4

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319055","number":319055,"commit":"70ed4bafc4353364bcdbcb3b95e2a3b35c9eeabd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318540","number":318540,"commit":"3138f96ea7b0f0b410e569a9a84e0f30de06517d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318023","number":318023,"commit":"36728a5e23c719d6c24c274aa05704eaee5817f3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317685","number":317685,"commit":"b5c76809e1c30e290f204c3b1696edeba364b5c4"}],"number":319055}
buildkite-pr-comment-->

florent-leborgne reviewed (commented):
Hi, docs output LGTM overall but will this endpoint be available for any versions, or from a specific stack version only?

If there is a version requirement, it should be specified with an `availability` attribute like this I believe (see [example](https://github.com/elastic/kibana/blob/main/x-pack/platform/plugins/shared/alerting/server/routes/maintenance_window/apis/external/update/update_maintenance_window_route.ts#L70-L71)):

```
options: {
        availability: {
          since: '9.1.0',
          stability: 'stable',
        },
```

CAWilson94 reviewed (approved):
Code LGTM!

Desk tested - enabled then disabled immediately after, status shows as disabled with error, null, as expected. 

Nice work 🚀

abhishekbhatia1710 committed:
Merge branch 'main' into ea-12990-disable-privmon

abhishekbhatia1710 committed:
Merge branch 'main' into ea-12990-disable-privmon

abhishekbhatia1710 committed:
Added the availability attribute for API docs

florent-leborgne reviewed (approved):
Docs LGTM

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/321424)
* Commit: 3df496b6e4fb9d949d8a655b0cd075ebb85ed62f

### Failed CI Steps
* [FTR Configs #99](https://buildkite.com/elastic/kibana-pull-request/builds/321424#01982d68-7ebe-4057-be4c-fb2998584780)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/321424#01982d68-7ebe-4057-be4c-fb2998584780) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/321424/jobs/01982d68-7ebe-4057-be4c-fb2998584780/artifacts/01982d8f-9ec3-42f4-8656-21fbcaf10498) FTR Configs #<span></span>99 / @<span></span>ess @<span></span>serverless SecuritySolution SIEM Migrations @<span></span>ess @<span></span>serverless @<span></span>serverlessQA Delete API "before each" hook for "should return 404 if migration ID does not exist"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :yellow_heart: [Build #319055](https://buildkite.com/elastic/kibana-pull-request/builds/319055) was flaky 70ed4bafc4353364bcdbcb3b95e2a3b35c9eeabd
* :green_heart: [Build #318540](https://buildkite.com/elastic/kibana-pull-request/builds/318540) succeeded 3138f96ea7b0f0b410e569a9a84e0f30de06517d
* :yellow_heart: [Build #318023](https://buildkite.com/elastic/kibana-pull-request/builds/318023) was flaky 36728a5e23c719d6c24c274aa05704eaee5817f3
* :broken_heart: [Build #317685](https://buildkite.com/elastic/kibana-pull-request/builds/317685) failed b5c76809e1c30e290f204c3b1696edeba364b5c4

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/321424","number":321424,"commit":"3df496b6e4fb9d949d8a655b0cd075ebb85ed62f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/319055","number":319055,"commit":"70ed4bafc4353364bcdbcb3b95e2a3b35c9eeabd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318540","number":318540,"commit":"3138f96ea7b0f0b410e569a9a84e0f30de06517d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318023","number":318023,"commit":"36728a5e23c719d6c24c274aa05704eaee5817f3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/317685","number":317685,"commit":"b5c76809e1c30e290f204c3b1696edeba364b5c4"}],"number":321424}
buildkite-pr-comment-->

machadoum reviewed (approved):
Could you please update the PR description? It doesn't match the implementation.

I desk-tested and left some minor comments.


* Please follow other API patterns for error handling. I left a comment in place for that.


* If you call the API before initialising the engine, you get the following response:
```json
{
  "message": "Saved object [privilege-monitoring-status/privilege-monitoring-default] not found",
  "status_code": 404
}
```
It would be better to get a more readable error. Something like "The engine was not initialised"


Double-checking: After the user disables the engine, is there no way to enable it again? Right?

abhishekbhatia1710 committed:
Merge branch 'main' into ea-12990-disable-privmon

abhishekbhatia1710 commented on x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/privilege_monitoring/privilege_monitoring_data_client.ts:
```diff
@@ -733,4 +736,67 @@ export class PrivilegeMonitoringDataClient {
+        error: null,
+      };
+    } catch (e) {
+      const msg = `Failed to disable Privileged Monitoring Engine: ${e.message}`;
+      this.log('error', msg);
+      errors.push(msg);
+
+      this.audit(
+        PrivilegeMonitoringEngineActions.DISABLE,
+        EngineComponentResourceEnum.privmon_engine,
+        'Failed to disable Privileged Monitoring Engine',
+        e
+      );
+      return {
+        status: PRIVILEGE_MONITORING_ENGINE_STATUS.STARTED,
+        error: errors,
+      };
```
Yes, the exception is already being captured and shown to the user but not in the `privMon_data_client.ts` but in `routes/disable.ts` file with the exact same implementation you described.

On error, the current implementation returns the required status code. For example, in this situation the exception was thrown and a 500 status error code was shown to the user. 

<img width="712" height="184" alt="image" src="https://github.com/user-attachments/assets/c4c2068b-7270-4762-92d7-3c69ee18d51f" />

abhishekbhatia1710 replied:
No, I was actually creating a new PR for this change, so I went ahead and resolved it. Anyway, the new PR for this change is : https://github.com/elastic/kibana/pull/229051

abhishekbhatia1710 commented:
> Double-checking: After the user disables the engine, is there no way to enable it again? Right?

@machadoum : If the user disables, user can enable it again via the `init` route. 

When the init is called after the disable, following logs are generated : 

```

[13:53:46.675] [DEBUG] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Initialized privileged monitoring engine saved object
[13:53:46.691] [ERROR] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Failed to create default index source for privilege monitoring: A monitoring entity source with the name "default-monitoring-index-default" already exists.
[13:53:46.691] [DEBUG] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Creating privilege user monitoring event.ingested pipeline
[13:53:46.694] [INFO ] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Privileged user monitoring ingest pipeline already exists.
[13:53:46.695] [INFO ] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Creating or updating index: .entity_analytics.monitoring.users-default
[13:53:46.700] [INFO ] [plugins.securitySolution] .entity_analytics.monitoring.users-default already exist
[13:53:46.706] [INFO ] [plugins.securitySolution] Updated mappings for .entity_analytics.monitoring.users-default
[13:53:46.712] [ERROR] [plugins.securitySolution] Failed to PUT settings for index .entity_analytics.monitoring.users-default: illegal_argument_exception
	Root causes:
		illegal_argument_exception: Can't update non dynamic settings [[index.mode]] for open indices [[.entity_analytics.monitoring.users-default/XQOnyc75Sxmpc83jtTd_cQ]] unless the `reopen` query parameter is set to true. Alternatively, close the indices, apply the settings changes, and reopen the indices
[13:53:46.712] [INFO ] [plugins.securitySolution] Generating Privmon API key
[13:53:48.070] [INFO ] [plugins.securitySolution] Scheduling privilege monitoring task with id entity_analytics:monitoring:privileges:engine:default:1.0.0
[13:53:49.965] [INFO ] [plugins.securitySolution] [Privilege Monitoring] Running privilege monitoring task
[13:53:49.977] [WARN ] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Index "entity_analytics.privileged_monitoring.default" not found — skipping.
[13:53:49.977] [DEBUG] [plugins.securitySolution] [Privileged Monitoring Engine][namespace: default] Found 0 stale users across all index sources.
[13:53:49.977] [INFO ] [plugins.securitySolution] [Privilege Monitoring] Finished running privilege monitoring task

```

abhishekbhatia1710 committed:
Adding the try catch for index source creation and check for pirvmon settings

hop-dev reviewed (approved):
Code review only!

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228961">](https://github.com/elastic/kibana/pull/228961)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
